### ISSUE TITLE
ceph-file-controller gets stuck in loops when having issues creating or deleting FS

### ISSUE THREAD

**Issue by psarossy** [rook/rook#5315]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
ceph-file-controller get stuck trying to create or delete filesystem forever

**Expected behavior:**
If ceph-file-controller can't create the filesystem it should back out and/or stop and provide some sort of actionable error message.

**How to reproduce it (minimal and precise):**

So I managed to reproduce this several times:

1. Deploy a new cluster based on the example configs (only change is that I enabled host networking).
1. Create a new FS with multiple pools in it (this documentation is pretty much non-existent btw), in my case one replicated and one erasure coded one. Make any sort of error in the config definition. The metadata and data0 pools are created but pool1 does not.
1. This results in errors like looping forever:
```
2020-04-23 05:33:18.680801 I | op-mon: parsing mon endpoints: a=10.0.0.11:6789,b=10.0.0.13:6789,c=10.0.0.14:6789
2020-04-23 05:33:19.344982 I | ceph-spec: adding finalizer "cephfilesystem.ceph.rook.io" on "myfs-demo"
2020-04-23 05:33:23.885510 I | ceph-file-controller: Creating filesystem myfs-demo
2020-04-23 05:33:27.858162 I | cephclient: creating replicated pool myfs-demo-metadata succeeded
2020-04-23 05:33:32.757954 I | cephclient: creating replicated pool myfs-demo-data0 succeeded
2020-04-23 05:33:34.343715 E | ceph-file-controller: failed to reconcile failed to create filesystem "myfs-demo": failed to create filesystem "myfs-demo": failed to create data pool "myfs-demo-data1": failed to create erasure code profile
 for pool "myfs-demo-data1": failed to set ec-profile: exit status 1
2020-04-23 05:33:36.196285 I | op-mon: parsing mon endpoints: a=10.0.0.11:6789,b=10.0.0.13:6789,c=10.0.0.14:6789
2020-04-23 05:33:41.343521 I | ceph-file-controller: Creating filesystem myfs-demo
2020-04-23 05:33:43.057092 E | ceph-file-controller: failed to reconcile failed to create filesystem "myfs-demo": failed to create filesystem "myfs-demo": failed to create data pool "myfs-demo-data1": failed to create erasure code profile
 for pool "myfs-demo-data1": failed to set ec-profile: exit status 1
```
1. Try to delete the pool.
1. This results in the following error message repeating forever
```
2020-04-23 05:35:35.015054 I | op-mon: parsing mon endpoints: a=10.0.0.11:6789,b=10.0.0.13:6789,c=10.0.0.14:6789
2020-04-23 05:35:35.889148 I | ceph-file-controller: Downing filesystem myfs-demo
2020-04-23 05:35:36.764528 E | ceph-file-controller: failed to reconcile failed to delete filesystem "myfs-demo". : failed to down filesystem "myfs-demo": failed to fail filesystem myfs-demo: exit status 2
```
1. Checking `ceph fs ls` show that there are no active filesystems defined
1. Check mate, can't recreate the FS with the correct config as the deletion is still in progress, and the deletion can't proceed as it can't down a filesystem that doesn't exist.

Alternatively, try to create 2 filesystems without enabling ROOK_ALLOW_MULTIPLE_FILESYSTEMS, then try to delete the second filesystem, this results in the same problem, loop forever that it can't create the FS, and deleting it fails also as it's trying to delete a filesystem that doesn't exist...

I'd appreciate any hints on how to make it forget about these, or clean them up. Every time something goes wrong with creating a pool or FS, it gets stuck in these loops and I can't find any documentation on how to deal with it, besides tearing down the cluster and rebuilding (which results in the namespace getting stuck in Terminating, requiring manual cleanup).

**File(s) to submit**:

All files are stock examples, except the filesystem config, which I attached.

**Environment**:
* OS (e.g. from /etc/os-release): Ubuntu 18.04.4 LTS 
* Kernel (e.g. `uname -a`): 4.15.0-96-generic
* Cloud provider or hardware configuration: baremetal
* Rook version (use `rook version` inside of a Rook Pod): v1.3.1
* Storage backend version (e.g. for ceph do `ceph -v`): 14.2.8
* Kubernetes version (use `kubectl version`): v1.18.2
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): none
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)): HEALTH_OK

**Comment by psarossy** [comment]
Seems github didn't like my attachment, so pasting it instead.

```
#################################################################################################################
# Create a filesystem with settings for a test environment where only a single OSD is required.
#  kubectl create -f filesystem-test.yaml
#################################################################################################################

apiVersion: ceph.rook.io/v1
kind: CephFilesystem
metadata:
  name: myfs-demo2
  namespace: rook-ceph
spec:
  metadataPool:
    replicated:
      size: 3
    deviceClass: ssd
  dataPools:
    - deviceClass: hdd
      failureDomain: host
      replicated:
        size: 3
    - deviceClass: hdd
      failureDomain: host
      erasureCoded:
        dataChunks: 3
        codingChunks: 2
  preservePoolsOnDelete: false
  metadataServer:
    activeCount: 2
    activeStandby: true
    placement:
      podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - rook-ceph-mds
            # topologyKey: kubernetes.io/hostname will place MDS across different hosts
            topologyKey: kubernetes.io/hostname
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 100
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - rook-ceph-mds
                # topologyKey: */zone can be used to spread MDS across different AZ
                # Use <topologyKey: failure-domain.beta.kubernetes.io/zone> in k8s cluster if your cluster is v1.16 or lower
                # Use <topologyKey: topology.kubernetes.io/zone>  in k8s cluster is v1.17 or upper
                topologyKey: topology.kubernetes.io/zone
```

**Comment by leseb** [comment]
What if you stop the operator so it stops reconciling, do some manual cleanup (just like you did) and then start the operator so that the CR can be deleted?

**Comment by psarossy** [comment]
So the issue for this case is that I'm not sure what to clean up, the operator never managed to create the resource it's trying to get rid of... Or it can't create the new item, which I can fix, but not sure how to stop it from trying to create it until then. Should a kubectl apply with the correct config just fix that (little hazy on how this part works, and didn't have much luck)?

In the filesystem example, it's trying to delete something that doesn't exist, and restarting the operator at least didn't do anything. I didn't actually clean anything up manually, it gets stuck on trying to down the filesystem after it removes the metadata and data0 pools, but then the FS no longer exists so there's nothing to down.
I did not remove the second definition from the config though when I sent it for deletion, should I be trying that just for the deletion? It's a bit of an odd scenario as I have to keep commenting out bits of the config for the deletion part and the error messages don't say what the issue is actually, so it's all manual inference.

I played around with just making the FS manually via the toolbox, but since the MDS-es are managed by the operator, they are not spun up for manual configs, so the cluster goes into HEALTH_WARN as there are <1 MDSes running, and it won't reconcile...

**Comment by psarossy** [comment]
I ended up brute force cleaning these up by editing the resource definitions, removing the finalizer and manually cleaning up the leftover ceph artifacts...

I do not think that approach is reasonable though as I have no idea what else rook does in the background to remove a filesystem or pool. This should be handled more gracefully...

**Comment by travisn** [comment]
@psarossy Are you able to create a filesystem with a single pool that is erasure coded? It's possible there is an issue with the filesystem EC pool since it's a less common scenario.
- When failing to "down" the filesystem, I believe #5342 will help with that

@leseb More ideas on why the CR can't be deleted or updated? With the controller runtime changes we should expect things to work more smoothly. Sounds like we are too aggressively returning errors instead of making a best effort to clean up and allow the deletion.

**Comment by psarossy** [comment]
@travisn Yes, I can create EC or replicated pools separately. This only surfaces when something does go with the creation. My hunch (didn't set up revision control for this file yet as I was testing) is that I had a typo of some sort in the second pool definition, which is why it got wedged.

The thing is, as I mentioned above, attempting to create the example FS twice with different names and ROOK_ALLOW_MULTIPLE_FILESYSTEMS also creates the same wedged state, where it can't create it (that's WAI), but you also cannot delete it as it can't down it, since it doesn't exist.

I don't think this is an issue with the actual FS creation in ceph, but more with how the orchestrator handles errors. There is no config validation from what I can tell (besides that it's valid yaml) so if anything goes wrong with the pool creation, this surfaces it seems.

I don't think #5342 will help with either of these scenarios, as it's not failing on MDS scaling, but on the fact that the filesystem definition no longer exists at all.

**Comment by NahaliY** [comment]
Hi, I have the same issue like @psarossy .
I can't create or delete FS.

Rook version: rook/ceph:v1.3.4
Ceph version: ceph/ceph:v14.2.9-20200414

`2020-05-01 16:36:46.373471 I | ceph-file-controller: Downing filesystem myfs-ec
2020-05-01 16:36:47.158500 E | ceph-file-controller: failed to reconcile failed to delete filesystem "myfs-ec". : failed to down filesystem "myfs-ec": failed to fail filesystem myfs-ec: exit status 2
`

**Comment by leseb** [comment]
@subhamkrai PTAL

**Comment by subhamkrai** [comment]
> @subhamkrai PTAL

this looks similar to #5796

**Comment by rajatsingh25aug** [comment]
Closed via #5796
